c10/core/DispatchKeySet.h

```diff
@@ -41,29 +121,127 @@ class DispatchKeySet final {
+      // For example, given DispatchKey::CPU, we should set:
+      // - the Dense functionality bit
+      // - the CPUBit backend bit
+      // first compute which bit to flip for the backend
+      auto backend_k = toBackendBit(k);
+      uint64_t backend_val = backend_k == BackendBit::InvalidBit ? 0 : 1ULL << (static_cast<uint8_t>(backend_k) - 1);
+      repr_ = functionality_val + backend_val;
+    } else {
+      // At this point, we should have covered every case except for alias keys.
+      // Technically it would be possible to add alias dispatch keys to a DispatchKeySet,
+      // but the semantics are a little confusing and this currently isn't needed anywhere.
+      repr_ = 0;
+    }
+
+  }
+
   explicit constexpr DispatchKeySet(std::initializer_list<DispatchKey> ks)
       : repr_(0) {
     for (auto k : ks) {
       repr_ |= DispatchKeySet(k).repr_;
     }
   }
+
+  constexpr uint64_t backend_bits_to_repr(std::initializer_list<BackendBit> ks) {
+    uint64_t repr = 0;
+    for (auto k : ks) {
+      repr |= DispatchKeySet(k).repr_;
+    }
+    return repr;
+  }
+
+  explicit constexpr DispatchKeySet(std::initializer_list<BackendBit> ks)
+      // Note: for some reason, putting this logic directly in the constructor
+      // appears to fail to compile on CUDA 10.1.
+      // See an example internal failure at
+      // https://www.internalfb.com/intern/skycastle/run/76561193669136035/artifact/actionlog.76561193742069401.stderr
+      : repr_(backend_bits_to_repr(ks)) {}
+
   // Test if a DispatchKey is in the set
-  bool inline has(DispatchKey t) const {
+  inline bool has(DispatchKey t) const {
     TORCH_INTERNAL_ASSERT_DEBUG_ONLY(t != DispatchKey::Undefined);
-    return static_cast<bool>(repr_ & DispatchKeySet(t).repr_);
+    return has_all(DispatchKeySet(t));
+  }
+  constexpr bool has(BackendBit t) const {
+    return has_all(DispatchKeySet(t));
+  }
+
+  // Test if a DispatchKey is in the set
+  // Given a DispatchKeySet of functionality keys and (potentially) backend keys,
+  // tests if all of them are in the current set.
+  constexpr bool has_all(DispatchKeySet ks) const {
+    return static_cast<bool>((repr_ & ks.repr_) == ks.repr_);
+  }
+
+  // Given a DispatchKeySet of functionality keys and (potentially) backend keys,
+  // tests if any of them are in the current set.
+  // This could technically be pretty easily implemented using has().
+  // It is strictly a perf optimization though.
+  // There are many places in the code base
+  // where we want to test for multiple functionality keys together.
+  // HOWEVER, runtime per-backend functionality keys aren't allowed to be used with this function,
+  // because you can end up with weird results.
+  // e.g. DispatchKeySet(DispatchKey::AutogradCPU).has_any(DispatchKeySet(DispatchKey::CPU)) would return true.
+  inline bool has_any(DispatchKeySet ks) const {
+    //TORCH_INTERNAL_ASSERT_DEBUG_ONLY(
```

Comment:
Yes :) I'll re-enable it before landing. I was paranoid that the assert might be running in my benchmarks since I was including debug symbols, but that probably isn't true (I think it only runs when `DEBUG=1` is set?)